File: backend/app/services/recipe_service.py
from ..data.repositories.recipe_repository import RecipeRepository
from ..data.schemas.food import FoodResponse
from ..data.schemas.recipe import RecipeCreate, RecipeIngredientDetail


class RecipeService:

    def __init__(self, repo: RecipeRepository):
        self.repo = repo

    async def create_recipe(self, data: RecipeCreate) -> FoodResponse:
        food = await self.repo.create(data)
        return FoodResponse.model_validate(food)

    async def get_ingredients(self, recipe_id: int) -> list[RecipeIngredientDetail]:
        ingredients = await self.repo.get_ingredients(recipe_id)
        return [
            RecipeIngredientDetail(
                food_id=ing.food_id,
                food_name=ing.food.name,
                quantity_grams=ing.quantity_grams,
                food_kcal=ing.food.kcal,
                food_protein=ing.food.protein,
                food_carbohydrates=ing.food.carbohydrates,
                food_fat=ing.food.fat,
            )
            for ing in ingredients
        ]

    async def update_recipe(self, recipe_id: int, data: RecipeCreate) -> FoodResponse:
        food = await self.repo.update(recipe_id, data)
        return FoodResponse.model_validate(food)

File: frontend/services/recipeService.ts
const API_URL = process.env.NEXT_PUBLIC_API_URL;

export interface RecipeIngredientIn {
  food_id: number;
  quantity_grams: number;
}

export interface RecipeCreate {
  name: string;
  description?: string;
  image_url?: string;
  ingredients: RecipeIngredientIn[];
}

export interface RecipeIngredientDetail {
  food_id: number;
  food_name: string;
  quantity_grams: number;
  food_kcal: number;
  food_protein: number;
  food_carbohydrates: number;
  food_fat: number;
}

export async function createRecipe(data: RecipeCreate): Promise<{ id: number }> {
  const res = await fetch(`${API_URL}/api/recipes`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  const json = await res.json().catch(() => ({}));
  if (!res.ok) throw new Error(json.detail || 'Eroare la salvarea rețetei.');
  return json;
}

export async function getRecipeIngredients(recipeId: number): Promise<RecipeIngredientDetail[]> {
  const res = await fetch(`${API_URL}/api/recipes/${recipeId}/ingredients`);
  const json = await res.json().catch(() => ([]));
  if (!res.ok) throw new Error('Eroare la încărcarea ingredientelor.');
  return json as RecipeIngredientDetail[];
}

export async function updateRecipe(recipeId: number, data: RecipeCreate): Promise<{ id: number }> {
  const res = await fetch(`${API_URL}/api/recipes/${recipeId}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  const json = await res.json().catch(() => ({}));
  if (!res.ok) throw new Error(json.detail || 'Eroare la actualizarea rețetei.');
  return json;
}
